### PR TITLE
fix plt.show doesn't warn if a non-GUI backend

### LINKED ISSUE #7908: plt.show doesn't warn if a non-GUI backend is being used
If a non-GUI background is being used, and `plt.gcf().show()` is called it raises the warning

```
/Users/dstansby/matplotlib/lib/matplotlib/figure.py:402: UserWarning: matplotlib is currently using a non-GUI backend, so cannot show the figure
  "matplotlib is currently using a non-GUI backend, "
```

However, calling `plt.show` does not raise this warning, and just silently doesn't show anything. I would expect at least the same warning to be raised.

### PR DESCRIPTION
**Pull request by zhoubecky**
## PR Summary
fix plt.show doesn't warn if a non-GUI backend is being used #7908
will add command and unit tests later

### CODE REVIEW

**Comment by jklymak** [comment]
Thanks for the PR! I just editied the title to be more descriptive and to exclude the issue number (though definitely include the issue # in the PR summary!)

File: lib/matplotlib/backend_bases.py [hunk 1]
```diff
@@ -175,20 +175,28 @@ def draw_if_interactive(cls):
...
         for manager in managers:
-            manager.show()
+            try:
+                manager.show()
+            except NonGuiException:
+                if warn:
+                    import warnings
+                    warnings.warn(
+                        "matplotlib is currently using a non-GUI backend, "
+                        "so cannot show the figure")
```
Comment (WeatherGod) [inline lib/matplotlib/backend_bases.py]: Might be nice to say which backend is being used

Comment (WeatherGod) [reply lib/matplotlib/backend_bases.py]: Also, what's the policy now for warnings versus logging warnings?

Comment (timhoffm) [reply lib/matplotlib/backend_bases.py]: According to the [logging howto](https://docs.python.org/3/howto/logging.html)

> warnings.warn() in library code if the issue is avoidable and the client application should be modified to eliminate the warning
>
> logging.warning() if there is nothing the client application can do about the situation, but the event should still be noted

The former is the case here. One should not try to call show (or expect that it does something) on a non-GUI backend. So, warnings.warn is correct.

Though, I don't think we need the `warn` parameter. Client applications can use the filter mechanism of the warnings module, if they really want to suppress the warning.

File: lib/matplotlib/backend_bases.py [hunk 2]
```diff
@@ -175,20 +175,28 @@ def draw_if_interactive(cls):
                 cls.trigger_manager_draw(manager)
 
     @classmethod
-    def show(cls, block=None):
+    def show(cls, block=None, warn=True):
```
Comment (timhoffm) [inline lib/matplotlib/backend_bases.py]: I don't think we need the `warn` parameter. Client applications can use the filter mechanism of the warnings module, if they really want to suppress the warning.

Comment (zhoubecky) [reply lib/matplotlib/backend_bases.py]: ok, I'll remove this parameter. Thanks.

**Comment by zhoubecky** [comment]
Hi @jklymak , I have removed the warning parameter and added a test. Is there anything else I should do to make this better?

File: lib/matplotlib/tests/test_backend_svg.py [hunk 3]
```diff
@@ -179,3 +179,22 @@ def psfont(*args, **kwargs):
     ax.text(0.5, 0.5, 'hello')
     with tempfile.TemporaryFile() as tmpfile, pytest.raises(ValueError):
         fig.savefig(tmpfile, format='svg')
+
+
+def test_NonGui_warning():
+    matplotlib.use('Agg')
+    import matplotlib.pyplot as plt
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_backend_svg.py]: Duplicate import.

File: lib/matplotlib/tests/test_backend_svg.py [hunk 4]
```diff
@@ -179,3 +179,22 @@ def psfont(*args, **kwargs):
...
+    plt.plot([1, 2, 3])
+    plt.savefig('myfig')
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_backend_svg.py]: Not needed?

File: lib/matplotlib/tests/test_backend_svg.py [hunk 5]
```diff
@@ -179,3 +179,22 @@ def psfont(*args, **kwargs):
     ax.text(0.5, 0.5, 'hello')
     with tempfile.TemporaryFile() as tmpfile, pytest.raises(ValueError):
         fig.savefig(tmpfile, format='svg')
+
+
+def test_NonGui_warning():
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_backend_svg.py]: I'm a bit confused; why does this test belong in the SVG backend test file? (The name should also not be CamelCase.)

File: lib/matplotlib/backend_bases.py [hunk 6]
```diff
@@ -182,13 +182,20 @@ def show(cls, block=None):
...
             return
         for manager in managers:
-            manager.show()
+            try:
+                manager.show()
+            except NonGuiException:
+                import warnings
+                warnings.warn(
+                    "matplotlib is currently using a non-GUI backend, "
+                    "so cannot show the figure")
```
Comment (QuLogic) [inline lib/matplotlib/backend_bases.py]: It looks like you missed @WeatherGod's suggestion:
> Might be nice to say which backend is being used

File: lib/matplotlib/tests/test_backend_svg.py [hunk 7]
```diff
@@ -179,3 +179,22 @@ def psfont(*args, **kwargs):
...
+    matplotlib.use('Agg')
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_backend_svg.py]: To force an alternate backend, you can mark the function with a decorator: `@pytest.mark.backend('Agg')`

File: lib/matplotlib/figure.py [hunk 8]
```diff
@@ -454,10 +455,10 @@ def show(self, warn=True):
             except NonGuiException:
                 pass
         if warn:
-            import warnings
             warnings.warn(
-                "matplotlib is currently using a non-GUI backend, "
-                "so cannot show the figure")
+                ('matplotlib is currently using %s, which is a ' +
```
Comment (jklymak) [inline lib/matplotlib/figure.py]: Can you fix the docstring typo while you are working on this?

Comment (zhoubecky) [reply lib/matplotlib/figure.py]: sorry, I don't find the typo. Which line is it?

Comment (jklymak) [reply lib/matplotlib/figure.py]: "warm" should be "warn"...  Unless my screen is blurry.

Comment (zhoubecky) [reply lib/matplotlib/figure.py]: Yes... I think so. It is warn. :)

Comment (jklymak) [reply lib/matplotlib/figure.py]: Line 434/435

Comment (zhoubecky) [reply lib/matplotlib/figure.py]: oh I see it. Will fix. Thank you

**Comment by zhoubecky** [comment]
Thank you @jklymak , I have updated docstring and squash all commits into one. Is there anything I should do?

**Comment by jkseppan** [comment]
I merged master into this branch to address the merge conflict, and now Github is showing this as a really huge patch (+3,415 −5,454). But on the command line this still shows up as the same change as before. I wonder if Github is confused or if I am mistaken...

**Comment by dstansby** [comment]
Rebasing on to master should fix things I think.

**Comment by jkseppan** [comment]
I think the Github diff machinery just has some limits for how complicated commit graphs it can handle. There were some old PRs (at least one from October 2017) merged recently, so if you look at `git log --graph` the history is a bit messy. The diffs look fine in the CLI.

@zhoubecky Thank you for your contribution!

**Pull request merged**